src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs

```diff
@@ -0,0 +1,346 @@
+                {
+                    goto Return;
+                }
+            }
+
+            idx = -1; // All characters are allowed.
+
+        Return:
+            return idx;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public override unsafe int FindFirstCharacterToEncodeUtf8(ReadOnlySpan<byte> utf8Text)
+        {
+            fixed (byte* ptr = utf8Text)
+            {
+                int idx = 0;
+
+#if NETCOREAPP
+                if (Sse2.IsSupported)
+                {
+                    sbyte* startingAddress = (sbyte*)ptr;
+                    while (utf8Text.Length - 16 >= idx)
+                    {
+                        Debug.Assert(startingAddress >= ptr && startingAddress <= (ptr + utf8Text.Length - 16));
+
+                        // Load the next 16 bytes.
+                        Vector128<sbyte> sourceValue = Sse2.LoadVector128(startingAddress);
+
+                        // Check if any of the 16 bytes need to be escaped.
+                        Vector128<sbyte> mask = Sse2Helper.CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin(sourceValue);
+
+                        int index = Sse2.MoveMask(mask.AsByte());
+                        // If index == 0, that means none of the 16 bytes needed to be escaped.
+                        // TrailingZeroCount is relatively expensive, avoid it if possible.
+                        if (index != 0)
+                        {
+                            // Found at least one byte that needs to be escaped, figure out the index of
+                            // the first one found that needed to be escaped within the 16 bytes.
+                            int tzc = BitOperations.TrailingZeroCount(index);
+                            Debug.Assert(tzc >= 0 && tzc <= 16);
+                            idx += tzc;
+                            goto Return;
+                        }
+                        idx += 16;
+                        startingAddress += 16;
+                    }
+
+                    // Process the remaining bytes.
+                    Debug.Assert(utf8Text.Length - idx < 16);
+                }
+#endif
+
+                for (; idx < utf8Text.Length; idx++)
+                {
+                    Debug.Assert((ptr + idx) <= (ptr + utf8Text.Length));
+                    if (NeedsEscaping(*(ptr + idx)))
+                    {
+                        goto Return;
+                    }
+                }
+
+                idx = -1; // All bytes are allowed.
+
+            Return:
+                return idx;
+            }
+        }
+
+        // The worst case encoding is 6 output chars per input char: [input] U+FFFF -> [output] "\uFFFF"
+        // We don't need to worry about astral code points since they're represented as encoded
+        // surrogate pairs in the output.
+        public override int MaxOutputCharactersPerInputCharacter => 12; // "\uFFFF\uFFFF" is the longest encoded form
```

Comment:
The code below is duplicated from https://github.com/dotnet/corefx/blob/6da4dd82bd6b33a1a2f8f56e43498fdceb2e41dc/src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/JavaScriptEncoder.cs#L71 (up to L315).

Derive this class from `DefaultJavaScriptEncoder`, so that the code can be reused (from the base-class)?

Reply:
I considered that, but `DefaultJavaScriptEncoder` is a sealed class and hence we can't derive from it.

Reply:
Removing `sealed` might not be an option (perf-wise), so maybe move the "shared" code to a helper and use from both places?

Reply:
I moved the shared code into a common static helper class. `TryEncodeUnicodeScalar` cannot be moved fully since it calls the virtual `WillEncode` which is different and extracting out the common snippets outside of that within `TryEncodeUnicodeScalar` doesn't seem worth it (especially since `UnsafeRelaxedJavaScriptEncoder.TryEncodeUnicodeScalar` has difference there too).